pkg/cluster/machine.go

```diff
@@ -63,9 +95,87 @@ func (m dockerMachine) CPUs(ctx context.Context) (int, error) {
 	return info.NCPU, nil
 }
 
-func (m dockerMachine) Apply(ctx context.Context, c *api.Cluster) (*api.Cluster, error) {
-	// TODO(nick): Implement this
-	return nil, fmt.Errorf("cluster type %s not configurable", c.Product)
+func (m dockerMachine) EnsureExists(ctx context.Context) error {
+	_, err := m.dockerClient.ServerVersion(ctx)
+	if err == nil {
+		return nil
+	}
+
+	klog.V(2).Infoln("No Docker daemon running. Attempting to start Docker.")
+	if m.os == "darwin" {
+		err := m.d4m.start(ctx)
+
+		dur := 60 * time.Second
+		_, _ = fmt.Fprintf(m.errOut, "Waiting %s for Docker Desktop to boot...\n", duration.ShortHumanDuration(dur))
+		err = wait.Poll(time.Second, dur, func() (bool, error) {
+			_, err := m.dockerClient.ServerVersion(ctx)
+			isSuccess := err == nil
+			return isSuccess, nil
+		})
+		if err != nil {
+			return fmt.Errorf("timed out waiting for Docker to start")
+		}
+		klog.V(2).Infoln("Docker started successfully")
+		return nil
+	}
+
+	if m.os == "windows" {
+		return fmt.Errorf("Please install Docker for Desktop: https://www.docker.com/products/docker-desktop")
+	}
+	return fmt.Errorf("Please install Docker for Linux: https://docs.docker.com/engine/install/")
+}
+
+func (m dockerMachine) Restart(ctx context.Context, desired, existing *api.Cluster) error {
+	canChangeCPUs := m.os == "darwin"
+	if existing.Status.CPUs < desired.MinCPUs && !canChangeCPUs {
```

Comment:
also worth checking "if desired product == DockerDesktop and k8s not enabled and os != darwin" and throwing an error for "can't start k8s for you"?

(the "k8s not enabled" check would probably need to be passed in from the existingCluster.Status stuff done when deciding whether to call `Restart`)

Reply:
i think that makes more sense in the cluster admin subsystem, because it has a kubectl client for checking to see if docker for windows is running OK. (docker-for-desktop is a weird case because it's both the machine driver and the cluster administrator)